test/org/javarosa/core/model/actions/SetLocationActionTest.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright 2019 Nafundi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.javarosa.core.model.actions;
+
+import org.javarosa.core.model.CoreModelModule;
+import org.javarosa.core.model.FormDef;
+import org.javarosa.core.model.data.DecimalData;
+import org.javarosa.core.model.data.IAnswerData;
+import org.javarosa.core.model.data.StringData;
+import org.javarosa.core.model.instance.InstanceInitializationFactory;
+import org.javarosa.core.model.instance.TreeReference;
+
+import org.javarosa.core.services.PrototypeManager;
+import org.javarosa.core.util.JavaRosaCoreModule;
+import org.javarosa.core.util.externalizable.DeserializationException;
+import org.javarosa.form.api.FormEntryController;
+import org.javarosa.form.api.FormEntryModel;
+import org.javarosa.model.xform.XFormsModule;
+import org.javarosa.xform.parse.XFormParseException;
+import org.junit.Test;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import static org.javarosa.core.util.externalizable.ExtUtil.defaultPrototypes;
+import static org.javarosa.test.utils.ResourcePathHelper.r;
+import static org.javarosa.xform.parse.FormParserHelper.parse;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNull;
+
+public class SetLocationActionTest {
+    @Test(expected = XFormParseException.class)
+    public void when_namespaceIsNotOdk_exceptionIsThrown() throws IOException {
+        parse(r("setlocation-action-bad-namespace.xml"));
+    }
+
+    @Test
+    public void when_instanceIsLoaded_locationIsSetAtTarget() throws IOException {
+        final FormDef formDef = parse(r("setlocation-action-instance-load.xml"));
+        formDef.initialize(true, new InstanceInitializationFactory());
+        TreeReference targetRef = createTargetRef(null);
+
+        assertEquals(SetLocationAction.NO_IMPLEMENTATION_MESSAGE, getValueAt(formDef, targetRef));
+    }
+
+    @Test
+    public void when_triggerNodeIsUpdated_locationIsSetAtTarget() throws IOException {
+        // Given
```

Comment:
Would you consider doing this?

```java
@Test
public void when_triggerNodeIsUpdated_locationIsSetAtTarget() {
    Scenario scenario = Scenario.init(r("setlocation-action-value-changed.xml"));

    // Answering a question triggers an "xforms-value-changed" event on it
    scenario.answer("/data/text", "some answer");

    assertThat(
        scenario.answerOf("/data/location"),
        is(stringAnswer(SetLocationAction.NO_IMPLEMENTATION_MESSAGE))
    );
}
```
(the `Scenario` class would only require a couple of small changes)